compiler/ast/src/main/kotlin/com/uber/xprocessing/ext/XType.kt

```diff
@@ -312,8 +312,16 @@ fun XType.isPrimitive(): Boolean {
 }
 
 private fun XType.hasCollectionType(): Boolean {
-  return this.typeElement?.name.orEmpty() in collectionTypes ||
-      typeArguments.any { it.hasCollectionType() }
+  val visited = mutableSetOf<XType>()
+  val queue = mutableListOf(this@hasCollectionType)
+  while (queue.isNotEmpty()) {
+    val type = queue.removeFirst()
+    if (type in visited) continue
```

Comment:
How big will visited get? This becomes an O(N^2) right? Do we move to something constant since this will effect build times?

Reply:
`visited` uses a `mutableSetOf` which uses `LinkedHashSet` behind the curtains. 
It's O(1) (constant) time complexity both for adding and for checking if set contains some element.

I believe the auxiliary list works the same as the previous recursive solution (difference is instead of leveraging the call stack, I'm using my own list), so no complexity degradation here as well.

So, all in all, I don't expect any build time degradation (I took that into consideration when designing the solution, but didn't bother to collect any metrics/benchmarking)